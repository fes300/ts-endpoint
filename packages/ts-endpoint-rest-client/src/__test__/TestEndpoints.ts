import * as t from 'io-ts';
import { Endpoint } from 'ts-endpoint';
import { RESTEndpoints } from '../types/RESTEndpoints';

const Actor = t.strict({
  id: t.string,
  name: t.string,
  createdAt: t.string,
  updatedAt: t.string,
}, 'Actor');

const TestEndpoints = {
  Actor: RESTEndpoints({
    Get: Endpoint({
      Method: 'GET',
      getPath: ({ id }) => `/actors/${id}`,
      Input: { Params: t.type({ id: t.string }), Query: undefined },
      Output: t.strict({ data: Actor }),
    }),
    List: Endpoint({
      Method: 'GET',
      getPath: () => `/actors`,
      Input: {
        Query: t.partial({
          _start: t.number,
          _end: t.number,
          ids: t.array(t.string),
        }),
      },
      Output: t.strict({ data: t.array(Actor), total: t.number }),
    }),
    Create: Endpoint({
      Method: 'POST',
      getPath: () => `/actors`,
      Input: { Body: Actor },
      Output: t.strict({ data: Actor }),
    }),
    Edit: Endpoint({
      Method: 'PUT',
      getPath: ({ id }) => `/actors/${id}`,
      Input: {
        Params: t.type({ id: t.string }),
        Body: Actor,
      },
      Output: t.strict({ data: Actor }),
    }),
    Delete: Endpoint({
      Method: 'DELETE',
      getPath: ({ id }) => `/actors/${id}`,
      Input: { Params: t.type({ id: t.string }) },
      Output: t.boolean,
    }),
    Custom: {
      GetSiblings: Endpoint({
        Method: 'GET',
        getPath: ({ id }) => `/actors/${id}/siblings`,
        Input: { Params: t.type({ id: t.string }) },
        Output: t.strict({ data: t.array(Actor) }),
      }),
    },
  }),
};

export { TestEndpoints };
