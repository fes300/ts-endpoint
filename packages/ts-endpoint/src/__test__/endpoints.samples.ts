import { Schema } from 'effect';
import { Endpoint } from '../Endpoint';

export const GetUserCrayons = Endpoint({
  Input: {
    Params: Schema.Struct({ id: Schema.Number }),
    Query: Schema.Struct({ color: Schema.String }),
  },
  Method: 'GET',
  getPath: ({ id }) => `users/${id.toString()}/crayons`,
  Output: Schema.Struct({ crayons: Schema.Array(Schema.String) }),
});

export const GetUserCrayon = Endpoint({
  Input: {
    Params: Schema.Struct({ userId: Schema.Number, crayonId: Schema.Number }),
    Query: Schema.Struct({ color: Schema.String }),
  },
  Method: 'GET',
  getPath: ({ userId, crayonId }) => `users/${userId.toString()}/crayons/${crayonId.toString()}`,
  Output: Schema.Struct({ crayons: Schema.Array(Schema.String) }),
});

export const ListCrayons = Endpoint({
  Input: {
    Query: Schema.Struct({ color: Schema.String }),
  },
  Method: 'GET',
  getPath: () => `users/crayons`,
  Output: Schema.Struct({ crayons: Schema.Array(Schema.String) }),
});

export const CreateCrayon = Endpoint({
  Input: {
    Body: Schema.Struct({ color: Schema.String }),
  },
  Method: 'POST',
  getPath: () => `users/crayons`,
  Output: Schema.Struct({ crayons: Schema.Array(Schema.String) }),
});

export const EditCrayon = Endpoint({
  Input: {
    Params: Schema.Struct({ id: Schema.Number }),
    Body: Schema.Struct({ color: Schema.String }),
  },
  Method: 'PUT',
  getPath: ({ id }) => `users/${id.toString()}/crayons`,
  Output: Schema.Struct({ crayons: Schema.Array(Schema.String) }),
});

export const DeleteCrayon = Endpoint({
  Input: {
    Params: Schema.Struct({ id: Schema.Number }),
  },
  Method: 'DELETE',
  getPath: ({ id }) => `users/${id.toString()}/crayons`,
  Output: Schema.Struct({ crayons: Schema.Array(Schema.String) }),
});
