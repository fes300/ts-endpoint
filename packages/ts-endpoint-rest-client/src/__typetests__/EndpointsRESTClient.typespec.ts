import { GetEndpointsRESTClient } from '../EndpointsRESTClient';
import { TestEndpoints } from '../__test__/TestEndpoints';

const Q = GetEndpointsRESTClient({} as any)(TestEndpoints)


// @dts-jest:fail:snap should not allow params with wrong type
Q.Actor.get({ id: 1 })

// @dts-jest:fail:snap should not allow call with empty params
Q.Actor.getList({ })
