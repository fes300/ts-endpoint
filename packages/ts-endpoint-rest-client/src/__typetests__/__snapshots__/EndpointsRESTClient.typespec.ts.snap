// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not allow call with empty params (type) should match snapshot 1`] = `
"Argument of type '{}' is not assignable to parameter of type 'Omit<GetListParams, "filter"> & { filter: Partial<{ _start?: number; _end?: number; ids?: string[]; }>; }'.
  Type '{}' is missing the following properties from type 'Omit<GetListParams, "filter">': pagination, sort"
`;

exports[`should not allow params with wrong type (type) should match snapshot 1`] = `"Type 'number' is not assignable to type 'string'."`;
